server/src/main/java/org/elasticsearch/action/admin/indices/settings/get/GetSettingsResponse.java

```diff
@@ -72,4 +84,65 @@ public void writeTo(StreamOutput out) throws IOException {
             Settings.writeSettingsToStream(cursor.value, out);
         }
     }
+
+    public static GetSettingsResponse fromXContent(XContentParser parser) throws IOException {
+        HashMap<String, Settings> indexToSettings = new HashMap<>();
+        if (parser.currentToken() == null) {
+            parser.nextToken();
+        }
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+        parser.nextToken();
+
+        while (!parser.isClosed() && parser.currentToken() == XContentParser.Token.FIELD_NAME) {
+            String indexName = parser.currentName();
+            parser.nextToken(); //go to per-index settings object
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+            XContentParserUtils.ensureFieldName(parser, parser.nextToken(), "settings");
+            parser.nextToken(); //go to settings object itself
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+            Settings settings = Settings.fromXContent(parser);
+            parser.nextToken(); //end per-index object
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.END_OBJECT, parser.currentToken(), parser::getTokenLocation);
+            parser.nextToken(); //we should now be positioned at the beginning of the next index's settings, or at the end
+            indexToSettings.put(indexName, settings);
+        }
+        return new GetSettingsResponse(ImmutableOpenMap.<String, Settings>builder().putAll(indexToSettings).build());
+    }
+    @Override
+    public String toString() {
+        return indexToSettings.toString();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
```

Comment:
we should rather take what we already have in `RestGetSettingsAction` to print out the response and move it here to the response, then also call it from there. You will probably be able to also move from `RestBuilderListener` to `RestToXContentListener`. I think that we are currently losing the include_defaults part...but looking deeper at it , I am not sure how that part will work out in the response given its dependencies. Leave a TODO for now I will look deeper next week.

Reply:
I had a deeper look at the settings filtering logic that we currently have in the REST layer, namely this part:

```
                    if (renderDefaults) {
                        builder.startObject("defaults");
                        settingsFilter.filter(indexScopedSettings.diff(cursor.value, settings)).toXContent(builder, request);
                        builder.endObject();
                    }
```

On one hand it is not a bad idea to keep the client code simple, you don't need filtering there as you just parse and print out what you get back from the server. On the other hand though I think we should have the rendering code in one place, rather than duplicated in the REST layer and in the response.

I am thinking that the filtering could be moved entirely to the transport layer here, based on the new includeDefaults flag that you are going to add to the request. In that case it will actually be read also from the transport client, and the response after such change will only contain the settings to be returned (and printed out) without needing an additional filtering pass. Let me know what you think about this.